node/src/template/single_use_tokens.rs

```diff
@@ -0,0 +1,200 @@
+use super::{actix::*, AssetTemplateContext, Contracts, Template, TemplateContext, TokenTemplateContext};
+use crate::{
+    db::models::{NewToken, Token, TokenStatus},
+    types::{Pubkey, TemplateID, TokenID},
+};
+use anyhow::{bail, Result};
+
+/// ***************** Asset contracts *******************
+
+//#[derive(Contracts)]
+pub enum AssetContracts {
+    //#[contract(issue_tokens)]
+    IssueTokens,
+}
+
+//#[contract(asset)]
+async fn issue_tokens<'a>(
+    context: AssetTemplateContext<'a>,
+    owner_pub_key: Pubkey,
+    token_ids: Vec<TokenID>,
+) -> Result<Vec<Token>>
+{
+    let mut tokens = Vec::with_capacity(token_ids.len());
+    let asset = &context.asset;
+    let new_token = move |token_id| NewToken {
+        token_id,
+        owner_pub_key: owner_pub_key.clone(),
+        asset_state_id: asset.id.clone(),
+        ..NewToken::default()
+    };
+    for data in token_ids.into_iter().map(new_token) {
+        if data.token_id.asset_id() != asset.asset_id {
```

Comment:
Thinking this over... does a token id really need the asset ID as part of it to guarantee uniqueness?

We have the asset id associated with the token records regardless and any call for instructions to the token include the asset id via the TokenCallParams so we can look it up that way. I think with that in mind any UUID passed in by the asset issuer would be fine as long as it's unique scoped down to that asset ID right?

We could still include the asset ID in the path for calling the instruction just wondering if we need to represent it in the db like this since we can derive it elsewhere thinking it over.

Reply:
Good point. Though apart of uniqueness (which uuid is enough) it provides extra routing benefits as we shall expect that TokenID and AssetID might to into the wild and end up on a node which do not have database. So this ensures that whoever received TokenID would be able to extract TemplateID, AssetID and Token uuid, that might be useful for locating master node for the asset and pulling information from relevant node then.

In regards to storage though - we can well store it as BIGINT + OID later in the DB. But that's optimization which we agreed to skip for now (but it will be really easy to do via FromSql ToSql)

Reply:
Sounds good